### ISSUE TITLE
Segfault during Kmeans for frame selection 

### ISSUE THREAD

**Issue by grahamfindlay** [DeepLabCut/DeepLabCut#1877]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Bug description

Similar issues #1651 and #1809 exist, but the proposed solutions fail in my case. 

Here's what I did:
1. When using the latest version of conda to create an environment using the `DEEPLABCUT.yaml` file provided, `wxpython` failed to build. Full stdout and stderr here: [conda_create_out.txt](https://github.com/DeepLabCut/DeepLabCut/files/8887086/conda_create_out.txt)
2. Following `setup.py`, I added  `wxpython<4.1` as a conda-supplied dependency to [DEEPLABCUT2.yaml](https://github.com/DeepLabCut/DeepLabCut/files/8887097/DEEPLABCUT2.txt). This solves the problem of `wxpython` not building. 
3. I get the OpenBLAS segfault during frame extraction, as reported by #1651 and #1809. The solution in #1651 was just to use the supplied environment file, which I am already doing. The solution in #1809 was to install scipy and scikit-learn with conda instead of pip. 
4. Following #1809, I added `scipy` and `scikit-learn` to [DEEPLABCUT3.txt](https://github.com/DeepLabCut/DeepLabCut/files/8887104/DEEPLABCUT3.txt). I hoped this would work, since I know that `conda-forge` often builds with different compiler options that better support CPUs with large numbers of cores. Full output from env creation [here](https://github.com/DeepLabCut/DeepLabCut/files/8887109/conda_create_out_3.txt). Sadly, I continued to get the same error, though now with a mangled stack trace and a SIGSEGV:
    ```
    Kmeans clustering ... (this might take a while)
    ...
    BLAS : Program is Terminated. Because you tried to allocate too many memory regions.
    *** Received signal 11 ***
    *** BEGIN MANGLED STACK TRACE ***
    Segmentation fault (core dumped)
    ```
    Output from  `python -m threadpoolctl --import sklearn` (also attached below, in the "Device type" section):
    ```
    [
      {
        "user_api": "blas",
        "internal_api": "openblas",
        "prefix": "libopenblas",
        "filepath": "/home/gfindlay/miniconda3/envs/datoviz/lib/python3.8/site-packages/numpy.libs/libopenblas64_p-r0-2f7c42d4.3.18.so",
        "version": "0.3.18",
        "threading_layer": "pthreads",
        "architecture": "SkylakeX",
        "num_threads": 64
      },
      {
        "user_api": "blas",
        "internal_api": "openblas",
        "prefix": "libopenblas",
        "filepath": "/home/gfindlay/.local/lib/python3.8/site-packages/scipy.libs/libopenblasp-r0-8b9e111f.3.17.so",
        "version": "0.3.17",
        "threading_layer": "pthreads",
        "architecture": "SkylakeX",
        "num_threads": 64
      },
      {
        "user_api": "openmp",
        "internal_api": "openmp",
        "prefix": "libgomp",
        "filepath": "/Volumes/scratch/neuropixels/t2_shared_envs/dlc_shared_t2/lib/libgomp.so.1.0.0",
        "version": null,
        "num_threads": 224
      }
    ]
    ```
    Note that the OpenMP `num_threads` is 224. 
5. #1809 referenced [this](https://github.com/scikit-learn/scikit-learn/issues/20539) scikit-learn issue, which further references [this](https://github.com/xianyi/OpenBLAS/issues/3321) OpenBLAS issue. See in particular [this](https://github.com/xianyi/OpenBLAS/issues/3321#issuecomment-885433142) comment. Since the `MiniBatchKMeans` that DLC uses does not offer an `n_jobs` kwargs like `DBSCAN` or `KNN` do, I set my `OPENBLAS_NUM_THREADS` and `OMP_NUM_THREADS` environment variables to 64. Didn't fix anything. In fact, `htop` sugggested that these variables were being ignored. I wondered if they might influence builds and not just runtime behavior (similar to `MAX_THREADS`), so I recreated the env. No luck. Perhaps importantly, there is no change to the output of `python -m threadpoolctl --import sklearn`. 
6. I took the extreme route and set `OPENBLAS_NUM_THREADS` and `OMP_NUM_THREADS` to 1. No luck. Again, no change to the output of `python -m threadpoolctl --import sklearn`. This makes me really wonder if the `OMP_NUM_THREADS` environment variable is being ignored. Maybe `omp_set_num_threads()` is overriding `OMP_NUM_THREADS`? I know this is possible, but usually is only done to _decrease_ the number of threads... I don't know of any other way to control OpenMP threading behavior. 

Unfortunately we have tried a containerized install as well, but cannot get the GUI to work on our remote servers using either rootless or rooted Docker. We are not yet clear on whether this is an issue with how the DLC container handles graphical displays or how Docker per se handles remote displays. Quite possibly the latter. Perhaps related to #1623. We will probably just give up and use local workstations for the GUI, and hope to at least get a containerized headless install working so we can train on the server GPU. 

Sorry I can't be more helpful! And sorry that this really is probably not an issue with DLC per se, but it seems to be affecting users, so I thought you might like more info. Thanks for DLC :

### Operating System

Ubunutu 20.04.4 LTS

### DeepLabCut version

2.2.1

### DeepLabCut mode

single animal

### Device type

I have an NVidia Tesla V100 32GB GPU, but I am attempting a conda-based CPU-only install on a 224-core Intel Xeon Platinum 8276M (Skylake, x86_64). For detailed info, see `lscpu` output [here](https://github.com/DeepLabCut/DeepLabCut/files/8887072/lscpu_out.txt).
 
As expected, `nproc` reports 224, `joblib.cpu_count()` reports 224, and `joblib.cpu_count(only_physical_cores=True)` reports 112. 

Here's the output of `python -m threadpoolctl --import sklearn`, which I suspect is highly relevant:
[threadpoolctl_out.txt](https://github.com/DeepLabCut/DeepLabCut/files/8887127/threadpoolctl_out.txt)

### Steps To Reproduce

See above. 

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/DeepLabCut/DeepLabCut/blob/master/CODE_OF_CONDUCT.md)

**Comment by grahamfindlay** [comment]
Not really relevant to the issue, but an update on my attempts to get a functional DLC workflow up and running, which maybe someone can help with: 

So, I was able to install DLC 2.2.1 on a Windows PC no problem, extract frames, label my data, etc. Then when I move my project to the Ubuntu server mentioned above and try to `deeplabcut.create_training_dataset()` in a working headless Docker container (DockerHub `deeplabcut:latest-core`), I get a number of errors that I believe might be due to a version discrepancy (v2.2.1 on Windows for labeling, v2.2.0.2 in the container), even with `windows2linux=True`. I could revert my desktop version, but would prefer to update my containerized version. The docs say, "For DeepLabCut 2.2.0.2 and onwards, we provide container containers on [DockerHub](https://hub.docker.com/r/deeplabcut/deeplabcut).", but It looks like my 2.2.0.2 is the latest, pushed 9 months ago. Any chance of 2.2.1 on DockerHub soon? Or should I attempt to generate my own container per [these](https://github.com/DeepLabCut/DeepLabCut/tree/master/docker) instructions?

**Comment by grahamfindlay** [comment]
Update: This seems to be dependent on the number of frames that need to be clustered. For example, when performing KMeans during `extract_outlier_frames()`, when there are typically far fewer frames that need clustering (say, 1000 vs. the 216,000 frames in a 60FPS 1h video), KMeans completes without error. 

We have still not successfully built a 2.2.1 Docker image, but were were able to get a mostly functional conda installation w/ GPU support and a GUI (just used for manual frame selection and labeling) using the following environment:
```
channels:
  - conda-forge
  - defaults
dependencies:
  - python=3.8
  - pip
  - jupyter
  - nb_conda
  - ffmpeg
  - scipy
  - scikit-learn
  - tensorflow-gpu
  - wxpython
  - pip:
    - "deeplabcut"
```
These [GPU-enabled tensorflow builds](https://conda-forge.org/blog/posts/2021-11-03-tensorflow-gpu/) make life very easy. Might be worth pointing to them in the install documentation.

**Comment by jeylau** [comment]
Hi @grahamfindlay, thank you very much for the very thorough report! I suggested the following in another related issue: https://github.com/DeepLabCut/DeepLabCut/issues/1826#issuecomment-1127457794
Could reducing cluster_resizewidth/increasing cluster_step help?
I regret I cannot help you much with the installation of scipy though; perhaps devs at their repo could shed some light on the issue.

**Comment by AlexEMG** [comment]
Indeed, for such long videos it is best to increase the cluster step (to sub-sample the video)!

**Comment by jeylau** [comment]
I'll close the issue, but feel free to reopen it if we can help you further (or to let us know how to fix a problematic scipy installation). ☺️
